src/renderer/components/QrReader/index.tsx

```diff
@@ -0,0 +1,457 @@
+  )
+
+  // General handler for errors which might occur during scanning.
+  const handleError = useCallback(
+    (error: any) => {
+      if (typeof error === 'string') {
+        onError(error)
+      } else {
+        onError(error.toString())
+      }
+
+      setError(true)
+    },
+    [onError]
+  )
+
+  // Read data from clipboard which potentially can be an image itself.
+  const handlePasteFromClipboard = useCallback(async () => {
+    try {
+      // Try interpreting the clipboard data as an image
+      const base64 = await runtime.readClipboardImage()
+      if (base64) {
+        const imageData = await base64ToImageData(base64)
+        const result = scanQrCode(
+          imageData.data,
+          imageData.width,
+          imageData.height
+        )
+        if (result) {
+          handleScanResult(result)
+          return
+        } else {
+          throw new Error('no data in clipboard image')
+        }
+      }
+
+      // .. otherwise return non-image data from clipboard directly
+      const data = await runtime.readClipboardText()
+      if (!data) {
+        throw new Error('no data in clipboard')
+      }
+      onScan(data)
+    } catch (error) {
+      userFeedback({
+        type: 'error',
+        text: `${tx('qrscan_failed')}: ${error}`,
+      })
+    }
+  }, [handleScanResult, onScan, tx, userFeedback])
+
+  // Read data from an external image file.
+  //
+  // We first trigger an "file open" dialog by automatically "clicking"
+  // an <input> element. The actual conversion and scanning of the file data
+  // happens afterwards.
+  const handleImportImage = useCallback(async () => {
+    if (inputRef.current) {
+      inputRef.current.click()
+    }
+  }, [])
+
+  const handleFileInputChange = useCallback(
+    async (event: React.ChangeEvent<HTMLInputElement>) => {
+      if (!event.target.files || event.target.files.length === 0) {
+        return
+      }
+      const file = event.target.files[0]
+
+      try {
+        // Convert file to correct image data and scan it
+        const base64 = await fileToBase64(file)
+        const imageData = await base64ToImageData(base64)
+        const result = scanQrCode(
+          imageData.data,
+          imageData.width,
+          imageData.height
+        )
+
+        if (result) {
+          handleScanResult(result)
+        } else {
+          userFeedback({
+            type: 'error',
+            text: `${tx('qrscan_failed')}: no data in image`,
+          })
+        }
+      } catch (error: any) {
+        handleError(error)
+      }
+
+      // Reset the input element again, otherwise we wouldn't be able to trigger
+      // another `onChange` again when selecting the same image
+      if (inputRef.current) {
+        inputRef.current.value = ''
+      }
+    },
+    [handleError, handleScanResult, tx, userFeedback]
+  )
+
+  // Show a context menu with different video input options to the user.
+  const handleSelectDevice = useCallback(
```

Comment:
name is confusing as it is not only about device selection, but also the other options